### PR TITLE
ARROW-1718: [C++/Python] Implement casts from timestamp to date32/64, properly handle NumPy datetime64[D] -> date32

### PR DESCRIPTION
**Pull request by wesm**
This was sort of a can of worms. cc @xhochy @cpcloud @BryanCutler

### CODE REVIEW

**Comment by BryanCutler** [comment]
+1 looks good, thanks for doing this!

File: .travis.yml [hunk 1]
```diff
@@ -51,12 +51,12 @@ matrix:
     os: linux
     group: deprecated
     before_script:
-    - export CC="gcc-4.9"
```
Comment (xhochy) [inline .travis.yml]: Is this expected to be in here?

Comment (wesm) [reply .travis.yml]: We aren't using clang-4.0 in any other build matrix, any objections to using that here (we use gcc 4.9 in other builds)?

Comment (xhochy) [reply .travis.yml]: We use clang-4.0 in the integration tests, but as we use gcc-4.9 in the manylinux1 builds I do not care much about the choice of compiler here as long as we test Python+old gcc together in some build.

**Review by xhochy — APPROVED** [review]
+1

File: .travis.yml (hunk 1 above)
Comment (wesm) [reply .travis.yml]: :+1:

**Pull request closed without merging**